File: day2/day2.rb
input = "forward 5
down 5
forward 8
up 3
down 8
forward 2"

input = File.read('day2.txt')
values = input.split("\n").map{ |row| row.split(" ") }.map{ |a, b| [a, b.to_i] }

pos = [0,0]
values.each do |command, value|
  case command
  when "forward" then pos[0] += value
  when "up" then pos[1] -= value
  when "down" then pos[1] += value
  end
end
puts pos.reduce(1,:*) # 1383564

# part 2
pos = [0,0]
aim = 0
values.each do |command, value|
  case command
  when "forward" then
    pos[0] += value
    pos[1] += value * aim
  when "up" then aim -= value
  when "down" then aim += value
  end
end
puts pos.reduce(1,:*) # 1488311643

File: day2/day2.dart
import 'dart:io';

void main() {
  var commands =
    File('day2.txt')
      .readAsLinesSync()
      .map((line) => Command.fromString(line))
      .toList();

  // part 1
  int position = 0;
  int depth = 0;
  for (Command command in commands) {
    switch (command.instruction) {
      case "forward":
        position += command.value;
        break;
      case "up":
        depth -= command.value;
        break;
      case "down":
        depth += command.value;
        break;
    }
  }
  print(position * depth);

  // part 2
  position = 0;
  depth = 0;
  int aim = 0;
  for (Command command in commands) {
    switch (command.instruction) {
      case "forward":
        position += command.value;
        depth += command.value * aim;
        break;
      case "up":
        aim -= command.value;
        break;
      case "down":
        aim += command.value;
        break;
    }
  }
  print(position * depth);
}


class Command {
  String instruction;
  int value;
  Command.fromString(String line) :
    instruction = line.split(' ')[0],
    value = int.parse(line.split(' ')[1]);
}
